src/storage/raw/raw_mvcc.rs

```diff
@@ -0,0 +1,191 @@
+    }
+
+    fn iter_cf(&self, cf: CfName, iter_opt: IterOptions) -> Result<Self::Iter> {
+        Ok(RawMvccIterator::new(self.snap.iter_cf(cf, iter_opt)?))
+    }
+
+    #[inline]
+    fn lower_bound(&self) -> Option<&[u8]> {
+        self.snap.lower_bound()
+    }
+
+    #[inline]
+    fn upper_bound(&self) -> Option<&[u8]> {
+        self.snap.upper_bound()
+    }
+
+    fn ext(&self) -> S::Ext<'_> {
+        self.snap.ext()
+    }
+}
+
+pub struct RawMvccIterator<I: Iterator> {
+    inner: I,
+    cur_key: Option<Vec<u8>>,
+    cur_value: Option<Vec<u8>>,
+    is_valid: Option<bool>,
+}
+
+impl<I: Iterator> RawMvccIterator<I> {
+    fn new(inner: I) -> Self {
+        RawMvccIterator {
+            inner,
+            cur_key: None,
+            cur_value: None,
+            is_valid: None,
+        }
+    }
+
+    fn update_cur_kv(&mut self, key: Vec<u8>, val: Vec<u8>) {
+        self.cur_key = Some(key);
+        self.cur_value = Some(val);
+        self.is_valid = Some(true);
+    }
+
+    fn clear_cur_kv(&mut self) {
+        self.cur_key = None;
+        self.cur_value = None;
+        self.is_valid = None;
+    }
+
+    fn move_to_prev_max_ts(&mut self, res: Result<bool>) -> Result<bool> {
+        if *res.as_ref().unwrap() && self.inner.valid()? {
+            self.update_cur_kv(self.inner.key().to_vec(), self.inner.value().to_vec());
+            self.inner.prev()?;
+        } else {
+            self.clear_cur_kv();
+            return Ok(false);
+        }
+        while *res.as_ref().unwrap() && self.inner.valid()? {
+            if self.cur_key.as_ref().unwrap() == self.inner.key() {
+                self.update_cur_kv(self.inner.key().to_vec(), self.inner.value().to_vec());
+                self.inner.prev()?;
+            } else {
+                break;
+            }
+        }
+        res
+    }
+}
+
+impl<I: Iterator> Iterator for RawMvccIterator<I> {
+    fn next(&mut self) -> Result<bool> {
+        let cur_key = self.inner.key().to_owned();
+        let mut res = self.inner.next();
+        while *res.as_ref().unwrap() && self.inner.valid()? && cur_key == self.inner.key() {
+            res = self.inner.next();
+        }
+        self.clear_cur_kv();
+        res
+    }
+
+    fn prev(&mut self) -> Result<bool> {
+        self.move_to_prev_max_ts(Ok(true))
+    }
+
+    fn seek(&mut self, key: &Key) -> Result<bool> {
+        self.inner.seek(key)
+    }
+
+    fn seek_for_prev(&mut self, key: &Key) -> Result<bool> {
+        let res = self.inner.seek_for_prev(key);
+        self.move_to_prev_max_ts(res)
+    }
+
+    fn seek_to_first(&mut self) -> Result<bool> {
+        self.inner.seek_to_first()
+    }
+
+    fn seek_to_last(&mut self) -> Result<bool> {
+        let res = self.inner.seek_to_last();
+        self.move_to_prev_max_ts(res)
+    }
+
+    fn valid(&self) -> Result<bool> {
+        match self.is_valid {
+            Some(v) => Ok(v),
+            None => self.inner.valid(),
+        }
```

Comment:
You can use [`unwrap_or`](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or) for this pattern.

Reply:
adressed